storage: avoid periodic reproposal of newly pending commands

petermattis opened the pull request:
Replica.tick() was previously reproposing any pending commands every
RaftElectionTimeoutTicks (1.5s). We now keep track of when a Raft
command was proposed and only repropose the command if it is older than
RaftElectionTimeoutTicks. This massively reduces unnecessary
reproposals.

See #8176.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8322)

<!-- Reviewable:end -->

petermattis committed:
storage: avoid periodic reproposal of newly pending commands

Replica.tick() was previously reproposing any pending commands every
RaftElectionTimeoutTicks (1.5s). We now keep track of when a Raft
command was proposed and only repropose the command if it is older than
RaftElectionTimeoutTicks. This massively reduces unnecessary
reproposals.

See #8176.

petermattis commented:
Unfortunately, this doesn't fix the performance problem in #8176. Before I add a test I want to get feedback on whether this is worthwhile to do and that this is the right approach.

bdarnell commented:
LGTM. A reasonable improvement even if it doesn't fix #8176. 

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/replica.go, line 1719 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KOISlEflc00OgjU9nfK:-KOISlEflc00OgjU9nfL:-1776085309) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ab392851050fe84e2e6ddc69de7ccb9291bca17/storage/replica.go#L1719)):*
> ```Go
> 		// RaftElectionTimeoutTicks is a reasonable approximation of how
> 		// long we should wait before deciding that our previous proposal
> 		// didn't go through.
> ```

Comment that the combination of the preceding `if` statement and the use of RaftElectionTimeoutTicks as the olderThanDelta argument to refreshPendingCmdsLocked means that commands will be refreshed when they have been pending for between 1 and 2 election timeouts. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-:-KOISYA85_9J1RBUfyoi:-357397362)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/replica.go, line 1719 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KOISlEflc00OgjU9nfK:-KOKJ3je_aIXEvD8Tgas:2047248762) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ab392851050fe84e2e6ddc69de7ccb9291bca17/storage/replica.go#L1719)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Comment that the combination of the preceding `if` statement and the use of RaftElectionTimeoutTicks as the olderThanDelta argument to refreshPendingCmdsLocked means that commands will be refreshed when they have been pending for between 1 and 2 election timeouts. 
</details>
We could have `Replica.tick()` always call `refreshPendingCmdsLocked`. Looking at this again with fresh morning eyes I think it is more appropriate for `refreshPendingCmdsLocked` to skip any command that is newer than the `olderThanTicks` value. The downside of calling `refreshPendingCmdsLocked` on every tick is the expense of looping over the pending commands, but that seems small.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322)*
<!-- Sent from Reviewable.io -->

petermattis committed:
storage: tweak TestReplicaBurstPendingCommandsAndRepropose

petermattis committed:
storage: s/reasonEmptyEntry/reasonNewLeaderOrConfigChange/g

petermattis commented:
Test added which revealed I wasn't setting `pendingCmd.proposedAtTicks` at the correct place. Two more commits added to fix issues found via inspection.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/replica.go, line 1719 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KOISlEflc00OgjU9nfK:-KOMadqMEg-ds6GlMSXy:-1587915995) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ab392851050fe84e2e6ddc69de7ccb9291bca17/storage/replica.go#L1719)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We could have `Replica.tick()` always call `refreshPendingCmdsLocked`. Looking at this again with fresh morning eyes I think it is more appropriate for `refreshPendingCmdsLocked` to skip any command that is newer than the `olderThanTicks` value. The downside of calling `refreshPendingCmdsLocked` on every tick is the expense of looping over the pending commands, but that seems small.
</details>
Looping over all pending commands on every tick doesn't seem great to me; we already need to reduce how much we do on each tick. We'd also need some code changes below so we only repropose every N ticks, instead of on every tick once the threshold is met. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-:-KOMcCc2UFuehZAf7wNV:1417088721)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
With the latest updates, does it help #8176?

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/replica.go, line 1719 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KOISlEflc00OgjU9nfK:-KON1lKARtUufAR6W2_g:-461137425) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ab392851050fe84e2e6ddc69de7ccb9291bca17/storage/replica.go#L1719)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Looping over all pending commands on every tick doesn't seem great to me; we already need to reduce how much we do on each tick. We'd also need some code changes below so we only repropose every N ticks, instead of on every tick once the threshold is met. 
</details>
Roger. I'll leave it as-is for now.

Note that we reset `pendingCmd.proposedAtTicks` every time a command is proposed. So we wouldn't be re-proposing a command on every tick after it passed the threshold.

Btw, I'm going to get rid of the `TODO` below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322)*
<!-- Sent from Reviewable.io -->

petermattis commented:
No, this didn't help with #8176. What it did almost completely eliminate are the `command observed at lease index` errors I was seeing. Prior to this change, a 1min run produced ~3000 of those errors. Afterwards I see 5-10. 

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-:-KON2JQEoOvypGBE7Le_:-1046344371)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1764 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON4Opv3MC0ryZ9QHAi:-KON4OpwscfgX01TPtTs:440481028) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2ea340045bd4347d1b48ec4a32b83ee5d1b92b6/storage/replica.go#L1764)):*
> ```Go
> 		if p.proposedAtTicks > olderThanTicks {
> 			// The command was proposed too recently, don't bother reproprosing or
> 			// reburbishing it yet. Note that if olderThanDelta is 0, olderThanTicks
> ```

refurbishing?

---

*[storage/replica.go, line 1752 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON51_iDeZPcTRBq39I:-KON51_iDeZPcTRBq39J:-693843438) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica.go#L1752)):*
> ```Go
> )
> 
> func (r *Replica) refreshPendingCmdsLocked(reason refreshRaftReason, olderThanDelta int) error {
> ```

s/olderThanTicks/reproposeAgeTicks/g?

---

*[storage/replica.go, line 1762 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON4q7i4Fli0JR5F-JU:-KON4q7i4Fli0JR5F-JV:1210058498) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica.go#L1762)):*
> ```Go
> 	// applied.
> 	maxWillRefurbish := r.mu.state.LeaseAppliedIndex // indexes <= will be refurbished
> 	olderThanTicks := r.mu.ticks - olderThanDelta
> ```

s/olderThanTicks/reproposeAtTicks/g?

---

*[storage/replica_test.go, line 6066 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON6kiQxbRRlyypJLOs:-KON6kiQxbRRlyypJLOt:-2027496748) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica_test.go#L6066)):*
> ```Go
> 	}
> 
> 	electionTicks := tc.store.ctx.RaftElectionTimeoutTicks
> ```

some commentary on the significance of this would be helpful (reading the test doesn't explain why this value matters for the purpose of reproposals).

there a comment below that is helpful ("Reproposals are only performed every electionTicks...") but something up here would help as well.

---

*[storage/replica_test.go, line 6068 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON6C6uFIhpJyr5eXKv:-KON6C6uFIhpJyr5eXKw:-472700760) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica_test.go#L6068)):*
> ```Go
> 	electionTicks := tc.store.ctx.RaftElectionTimeoutTicks
> 	for i := 0; i < 2*electionTicks; i++ {
> 		// Each iteration add another pending command.
> ```

adds?

---

*[storage/replica_test.go, line 6092 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON6wE-1ARLymqn4Jq3:-KON6wE-1ARLymqn4Jq4:-814690404) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica_test.go#L6092)):*
> ```Go
> 		}
> 
> 		// Check that only a single command was reproposed.
> ```

i don't think that's what this does.

---

*[storage/replica_test.go, line 6006 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON9RXgXHKn4g5WGZnT:-KON9RXhR-U9VBp3GDwW:304378936) ([raw file](https://github.com/cockroachdb/cockroach/blob/dab4e0514078ef84744ca7a111c51274ab3fdf9b/storage/replica_test.go#L6006)):*
> ```Go
> 			cmd := tc.rng.prepareRaftCommandLocked(ctx, makeIDKey(), repDesc, ba)
> 			tc.rng.insertRaftCommandLocked(cmd)
> 			chs = append(chs, cmd.done)
> ```

what's this about? commit message is unhelpful

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322)*
<!-- Sent from Reviewable.io -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=8322) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=8322) before we can accept your contribution.<br/>

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1764 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON4Opv3MC0ryZ9QHAi:-KOPjE3kGT8Oj7_hLxp9:b4cne8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2ea340045bd4347d1b48ec4a32b83ee5d1b92b6/storage/replica.go#L1764)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> refurbishing?
</details>
My fingers can't spell repropose or refurbish. Done.

---

*[storage/replica.go, line 1752 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON51_iDeZPcTRBq39I:-KOPkY0A8EDqcJMvoMvt:buolhfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica.go#L1752)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/olderThanTicks/reproposeAgeTicks/g?
</details>
Used `refreshAt{Delta,Ticks}` as the parameter controls both reproposing and refurbishing.

---

*[storage/replica.go, line 1762 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON4q7i4Fli0JR5F-JU:-KOPkcYcI56-hirYl0zi:b-d96vr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica.go#L1762)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/olderThanTicks/reproposeAtTicks/g?
</details>
Done as `refreshAtTicks`.

---

*[storage/replica_test.go, line 6066 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON6kiQxbRRlyypJLOs:-KOPjqzvbarzJ7w_7WeF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica_test.go#L6066)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> some commentary on the significance of this would be helpful (reading the test doesn't explain why this value matters for the purpose of reproposals).
> 
> there a comment below that is helpful ("Reproposals are only performed every electionTicks...") but something up here would help as well.
</details>
Done.

---

*[storage/replica_test.go, line 6068 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON6C6uFIhpJyr5eXKv:-KOPjm0uQQYl3tiTxSTA:b-rjtc3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica_test.go#L6068)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> adds?
</details>
I think it can go either way. I've rephrased.

---

*[storage/replica_test.go, line 6092 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON6wE-1ARLymqn4Jq3:-KOPjxsvlnCl80K7WUo2:ba8wku8) ([raw file](https://github.com/cockroachdb/cockroach/blob/90500489060d6be804d7c8a6082f21fa687f85dd/storage/replica_test.go#L6092)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i don't think that's what this does.
</details>
It originally did and then I changed the test. Changed to: `Gather up the reproprosed commands.`

---

*[storage/replica_test.go, line 6006 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322#-KON9RXgXHKn4g5WGZnT:-KOPjQBgL9o8LMANkeMd:b-dng0rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dab4e0514078ef84744ca7a111c51274ab3fdf9b/storage/replica_test.go#L6006)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's this about? commit message is unhelpful
</details>
The test was creating a `chs` slice and never appending to it. I'm not quite sure what the intention of the test was, just noticed this in passing. The test seems to pass both with and without this change. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8322)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on storage/replica.go:
```diff
@@ -1708,12 +1714,16 @@ func (r *Replica) tick() error {
+		// should wait before deciding that our previous proposal didn't go
+		// through. Note that the combination of the above condition and passing
+		// RaftElectionTimeoutTicks to refreshPendingCmdsLocked means that commands
+		// will be refreshed when they have been pending for 1 to 2 electionc
```
s/electionc/election/
